src/core/layout/qgslayoutitemmap.cpp

```diff
@@ -1893,23 +1893,27 @@ QList<QgsMapLayer *> QgsLayoutItemMap::layersToRender( const QgsExpressionContex
     scopedContext = createExpressionContext();
   const QgsExpressionContext *evalContext = context ? context : &scopedContext;
 
-  QList<QgsMapLayer *> renderLayers;
+  QList<QgsMapLayer *> renderLayers, highlightLayers;
+
+  if ( !layers().isEmpty() )
+    renderLayers = layers();
+  else
+    renderLayers = mLayout->project()->mapThemeCollection()->masterVisibleLayers();
+
+  // preserve the list of highlight layers to add if presetName is set
+  for ( const auto mlayer : renderLayers )
+  {
+    if ( mlayer->name().contains( "highlight_" ) )
```

Comment:
This is quite hacky. I think it should be handled in server instead, by modifying the theme definition to include all desired map layers including the highlight ones.

Reply:
Fully agree! But I was not able to find a more elegant solution (yet).

The same theme can be used in more than one map in the layout; and the highlight can be only on one of those maps. I think we can not change the theme.

I'm creating a more complicated combination to test this (a layout with several maps and several highlights, and some with layers following a theme).

Reply:
What about copying the theme, modifying it as desired, and then setting the map to follow the new theme?

Reply:
@nyalldawson I need some help, sorry! 
The highlight layers are already created on the server logic, as the other layers passed as parameters, before we know if the map item follows a project theme.
Here (at `QgsLayoutItemMap::layersToRender`) we were discarding all server layers (including the highlight ones) and use only the ones from the theme.
I've just added a few lines to preserve the highlighted layers, before dropping them all, if the map item follows a theme. It's a kind of a hack (not a proper refactor), I agree, but it does seem dangerous.
In summary, I need your help. I'm not seeing the overall picture as you do.

Reply:
What I'm suggesting is something like (and I don't know the internals of the server code to give specifics, sorry!)

- create the highlight layers
- before exporting the layout, you see which maps need to include the highlight layers
- for these, if any have a theme set, you copy the theme, add the highlight layers to it, and then set the map item to follow that new duplicate theme
- drop the temporary theme

Reply:
I agree with Nyall's comments, it should clearly be managed on server side. And the idea of creating a new temporary theme seems the way to go indeed.